topostats/scars.py

```diff
@@ -0,0 +1,470 @@
+        direction: str,
+        threshold_low: float,
+        threshold_high: float,
+        max_scar_width: int,
+        min_scar_length: int,
+    ) -> np.ndarray:
+        """
+        Mark scars within an image, returning a boolean 2D np.ndarray of pixels that have been detected as scars.
+
+        Parameters
+        ----------
+        img: np.ndarray
+            A 2-D image of the data to detect scars in.
+        direction: str
+            Options: 'positive', 'negative'. The direction of scars to detect. For example, to detect scars
+            that lie above the data, select 'positive'.
+        threshold_low: float
+            A floating point value, that when multiplied by the standard deviation of the image, acts as a
+            threshold for sharp inclines or descents in pixel values and thus marks potential scars.
+            A lower value will make the algorithm
+            more sensitive, and a higher value will make it less sensitive.
+        threshold_high: float
+            A floating point value that is used similarly to threshold_low, however sharp inclines or descents
+            that result in values in the mask higher than this threshold are automatically considered scars.
+        max_scar_width: int
+            An integer that restricts the algorithm to only mark scars that are as thin or thinner than this width.
+            This is important for ensuring that legitimate grain data is not detected as scarred data.
+            Note that width here is vertical, as scars are thin, horizontal features.
+        min_scar_length: int
+            An integer that restricts the algorithm to only mark scars that are as long or longer than this length.
+            This is important for ensuring that noise or legitimate but sharp datapoints do not get detected as scars.
+            Note that length here is horizontal, as scars are thin, horizontal features.
+        Returns
+        -------
+        marked: np.ndarray
+            Returns a 2-D image of the same shape as the data image, where each pixel's value represents
+            a metric for how strongly that pixel is considered to be a scar. Higher values mean more likely
+            to be a scar.
+        """
+
+        img = np.copy(img)
+
+        stddev = np.std(img)
+        marked = np.zeros(img.shape)
+
+        if direction == "positive":
+            # Find positive scars
+            for row in range(img.shape[0] - 1):
+                for col in range(img.shape[1]):
+
+                    self.mark_if_positive_scar(
+                        row_col=(row, col),
+                        stddev=stddev,
+                        img=img,
+                        marked=marked,
+                        threshold_low=threshold_low,
+                        max_scar_width=max_scar_width,
+                    )
+
+        elif direction == "negative":
+            # Find negative scars
+            for row in range(img.shape[0] - 1):
```

Comment:
You could perhaps reduce code duplication by having the conditional `if: ... elif:` within the loop as you are looping over the same range of `row`/`col` regardless of whether the `direction` is `positive` or `negative.

```
for row in range(img.shape[0] - 1):
    for col in range(img.shape[1]):
        if direction == "positive:
            self.mark_if_positive_scar(...)
        elif direction == "negative":
            self.mark_if_negative_scar(...)
        else:
            raise ValueError(f"Invalid direction : {direction}")
```

Reply:
Done 👍